### ISSUE TITLE
100x slow down with eager execution in tensorflow 2.0

### ISSUE THREAD

**Issue by Char-Aznable** [tensorflow/probability#629]
I am trying to fit an autoregressive flow to a unnormalized log-prob function. The input and output of the flow are both 9 dimensional. I originally got an implementation in tensorflow 1.0. Now I am switching to tensorflow 2.0, where I observe a 100x slow down when using eager execution as compared to a run with eager execution disabled. I am using tensorflow 2.0 compiled from source with GPU support on the same machine that is running the codes below. Also, I monitor the GPU utilization via `nvidia-smi` and see a very low GPU utilization with the eager execution. Here are my codes:

With eager execution (this version is about 100x slower):
```python
import numpy as np
import tensorflow as tf
import tensorflow_probability as tfp
tfd = tfp.distributions
tfb = tfp.bijectors
tf.random.set_seed(0)
np.random.seed(0)
```

<details><summary>model construction codes that are common to both version</summary>
<p>

```python
class Monomers:

    def __init__(self, beta=1.0, n=3, d0=1.0, kd=1.0,
                 boundCenter=np.array([0, 0, 0], dtype=float),
                 bound0=10.0, kbound=1.0):

        if boundCenter.size != 3:
            raise ValueError("Input boundary center is not array of 3")

        self.beta = 1.0
        self.n = 3
        self.nPairs = int(n*(n-1) / 2)
        self.d0 = d0
        self.d2cutoff = self.d0 * self.d0
        self.kd = 0.5 * kd
        self.boundCenter = tf.constant(boundCenter, dtype=tf.float32)
        self.bound0 = bound0
        self.bound02 = self.bound0 * self.bound0
        self.kbound = 0.5 * kbound

    @tf.function
    def dist2s(self, x):
        """Compute pair-wise squared distances among each of n monomers in
        the batch

        Args:
                x: `Tensor` of shape (..., n, 3). A batch of n monomers'
                3D positions.

        Returns:
            `Tensor` of shape (..., n, n) which are the distances between
            each n by n pairs of input positions, where `...` is the
            batch dimensions
        """
        return tf.reduce_sum(input_tensor=tf.math.squared_difference(x[..., tf.newaxis, :, :],
                                                   x[..., :, tf.newaxis, :]),
                             axis=-1)

    @tf.function
    def getUniquePair(self, m):
        """Reshape the input pairwise matrix so that only its unique elements
        are returned as a 1D array

        Args:
                m: `Tensor` of shape (..., self.n, self.n): a batch of
                pairwise matrices for each pair of particle

        Returns: `Tensor` of shape (..., self.nPairs): the unique elements in m
        """
        # get the unique pairwise matrix by masking the upper triangle
        # of the interdistance matrix
        ones = tf.ones_like(m, dtype=tf.bool)
        mask = tf.math.logical_xor(tf.linalg.band_part(ones, 0, -1),
                              tf.linalg.band_part(ones, 0, 0))
        # reshape to the input batch shape -- this should be of shape:
        # batch-shape x self.nPairs
        mUniqueFlat = tf.reshape(tf.boolean_mask(tensor=m, mask=mask),
                                 tf.concat([tf.shape(input=m)[:-2], [-1]],
                                           axis=0))
        return mUniqueFlat

    @tf.function
    def energyPair(self, x):
        """Compute the energy of pairwise interaction given a
        batch of configurations

        Args:
                x: `Tensor` of shape (..., self.n, 3) which is a batch of
                n-mer's 3D coordinates

        Returns:
            `Tensor` of shape (...) which is the a batch of pairwise
            interaction energy
        """
        # compute the pairwise squared distances
        d2s = self.dist2s(x)
        d2sUniqueFlat = self.getUniquePair(d2s)
        e = tf.reduce_sum(input_tensor=tf.compat.v2.where(
            d2sUniqueFlat < self.d2cutoff,
            self.kd * tf.square(tf.sqrt(d2sUniqueFlat) - self.d0),
            0), axis=-1)
        return e

    @tf.function
    def d2Bound(self, x):
        """Compute the distance squared to the center of the boundary condition

        Args:
                x: `Tensor` of shape (..., self.n, 3) which is a batch of
                n-mer's 3D coordinates

        Returns:
            `Tensor` of shape (..., self.n) which is the a batch of
            boundary energy
        """
        return tf.reduce_sum(input_tensor=tf.math.squared_difference(x, self.boundCenter),
                             axis=-1)

    @tf.function
    def energyBound(self, x):
        """Compute the energy due to boundary condition

        Args:
                x: `Tensor` of shape (..., self.n, 3) which is a batch of
                n-mer's 3D coordinates

        Returns:
            `Tensor` of shape (...) which is the a batch of boundary energy
        """
        d2s = self.d2Bound(x)
        e = tf.reduce_sum(
            input_tensor=tf.compat.v2.where(d2s > self.bound02,
                        self.kbound * tf.square(tf.sqrt(d2s) - self.bound0),
                        0),
            axis=-1)
        return e

    @tf.function
    def log_prob(self, x):
        """Compute the log-probability of a given batch of configurations

        Args:
                x: `Tensor` of shape (..., self.n, 3) which is a batch of
                n-mer's 3D coordinates

        Returns:
            `Tensor` of shape (...) which is the a batch of log-probability
        """
        return -self.beta * (self.energyPair(x) + self.energyBound(x))


"""thermodynamic temperature"""
T = 300
"""Boltzmann constant"""
kB = 0.0083144621
beta = 1.0 / (kB*T)
nMer = 3
nDOF = nMer * 3
system = Monomers(beta, nMer, 1.0, 10.0, np.zeros(3, dtype=float), 10.0, 10.0)
nBijectors = 5
bPermute = True
bijectors = []
permutations = []
"""
In order to fully permute all the dimension, I need to flatten out the
inner dimension so that tfb.Permute can work on it in MAF then I need
to reshape it for the log prob
"""
bijectors.append(tfb.Reshape(event_shape_in=[nMer, 3],
                             event_shape_out=[nDOF]))
for i in range(nBijectors):
    bijectors.append(tfb.Invert(tfb.MaskedAutoregressiveFlow(
        shift_and_log_scale_fn=tfb.masked_autoregressive_default_template(
            activation=tf.nn.sigmoid,
            hidden_layers=[nDOF]),
        unroll_loop=True)))
    if bPermute and i < nBijectors - 1:
        """Last bijector must not be permutation"""
        permutation = np.random.permutation(nDOF)
        bijectors.append(tfb.Permute(permutation=permutation, axis=-1))
        permutations.append(permutation)
bijectors.append(tfb.Reshape(event_shape_in=[nDOF], event_shape_out=[nMer, 3]))
"""The bijectors are applied from last to first -- need to reverse"""
flow = tfb.Chain(list(reversed(bijectors)))
baseDist = tfd.Independent(
    distribution=tfd.Uniform(low=tf.zeros([nMer, 3], dtype=tf.float32),
                             high=tf.ones([nMer, 3], dtype=tf.float32)),
    reinterpreted_batch_ndims=2)
dist = tfd.TransformedDistribution(
    distribution=baseDist,
    bijector=flow)
```
</p>
</details>

```python
optimizerReverse = tf.optimizers.Adam(1e-2)
nStepTrainingReverse = int(2e4)
stepsReverse = []
iStep = 0
for i in range(nStepTrainingReverse):
    stepsReverse.append(iStep)
    optimizerReverse.minimize(ELBOLoss, dist.trainable_variables)
    if i % 1000 == 0:
        print("Reverse: %10d\n" % (i))
    iStep += 1

```
Without eager execution:
```python
import numpy as np
import tensorflow as tf
import tensorflow_probability as tfp
tf.compat.v1.disable_eager_execution()
tf.compat.v1.reset_default_graph()
sess = tf.compat.v1.InteractiveSession()
tfd = tfp.distributions
tfb = tfp.bijectors
tf.compat.v1.set_random_seed(0)
np.random.seed(0)
```
<details><summary>model construction codes that are common to both version</summary>
<p>

```python
class Monomers:

    def __init__(self, beta=1.0, n=3, d0=1.0, kd=1.0,
                 boundCenter=np.array([0, 0, 0], dtype=float),
                 bound0=10.0, kbound=1.0):

        if boundCenter.size != 3:
            raise ValueError("Input boundary center is not array of 3")

        self.beta = 1.0
        self.n = 3
        self.nPairs = int(n*(n-1) / 2)
        self.d0 = d0
        self.d2cutoff = self.d0 * self.d0
        self.kd = 0.5 * kd
        self.boundCenter = tf.constant(boundCenter, dtype=tf.float32)
        self.bound0 = bound0
        self.bound02 = self.bound0 * self.bound0
        self.kbound = 0.5 * kbound

    @tf.function
    def dist2s(self, x):
        """Compute pair-wise squared distances among each of n monomers in
        the batch

        Args:
                x: `Tensor` of shape (..., n, 3). A batch of n monomers'
                3D positions.

        Returns:
            `Tensor` of shape (..., n, n) which are the distances between
            each n by n pairs of input positions, where `...` is the
            batch dimensions
        """
        return tf.reduce_sum(input_tensor=tf.math.squared_difference(x[..., tf.newaxis, :, :],
                                                   x[..., :, tf.newaxis, :]),
                             axis=-1)

    @tf.function
    def getUniquePair(self, m):
        """Reshape the input pairwise matrix so that only its unique elements
        are returned as a 1D array

        Args:
                m: `Tensor` of shape (..., self.n, self.n): a batch of
                pairwise matrices for each pair of particle

        Returns: `Tensor` of shape (..., self.nPairs): the unique elements in m
        """
        # get the unique pairwise matrix by masking the upper triangle
        # of the interdistance matrix
        ones = tf.ones_like(m, dtype=tf.bool)
        mask = tf.math.logical_xor(tf.linalg.band_part(ones, 0, -1),
                              tf.linalg.band_part(ones, 0, 0))
        # reshape to the input batch shape -- this should be of shape:
        # batch-shape x self.nPairs
        mUniqueFlat = tf.reshape(tf.boolean_mask(tensor=m, mask=mask),
                                 tf.concat([tf.shape(input=m)[:-2], [-1]],
                                           axis=0))
        return mUniqueFlat

    @tf.function
    def energyPair(self, x):
        """Compute the energy of pairwise interaction given a
        batch of configurations

        Args:
                x: `Tensor` of shape (..., self.n, 3) which is a batch of
                n-mer's 3D coordinates

        Returns:
            `Tensor` of shape (...) which is the a batch of pairwise
            interaction energy
        """
        # compute the pairwise squared distances
        d2s = self.dist2s(x)
        d2sUniqueFlat = self.getUniquePair(d2s)
        e = tf.reduce_sum(input_tensor=tf.compat.v2.where(
            d2sUniqueFlat < self.d2cutoff,
            self.kd * tf.square(tf.sqrt(d2sUniqueFlat) - self.d0),
            0), axis=-1)
        return e

    @tf.function
    def d2Bound(self, x):
        """Compute the distance squared to the center of the boundary condition

        Args:
                x: `Tensor` of shape (..., self.n, 3) which is a batch of
                n-mer's 3D coordinates

        Returns:
            `Tensor` of shape (..., self.n) which is the a batch of
            boundary energy
        """
        return tf.reduce_sum(input_tensor=tf.math.squared_difference(x, self.boundCenter),
                             axis=-1)

    @tf.function
    def energyBound(self, x):
        """Compute the energy due to boundary condition

        Args:
                x: `Tensor` of shape (..., self.n, 3) which is a batch of
                n-mer's 3D coordinates

        Returns:
            `Tensor` of shape (...) which is the a batch of boundary energy
        """
        d2s = self.d2Bound(x)
        e = tf.reduce_sum(
            input_tensor=tf.compat.v2.where(d2s > self.bound02,
                        self.kbound * tf.square(tf.sqrt(d2s) - self.bound0),
                        0),
            axis=-1)
        return e

    @tf.function
    def log_prob(self, x):
        """Compute the log-probability of a given batch of configurations

        Args:
                x: `Tensor` of shape (..., self.n, 3) which is a batch of
                n-mer's 3D coordinates

        Returns:
            `Tensor` of shape (...) which is the a batch of log-probability
        """
        return -self.beta * (self.energyPair(x) + self.energyBound(x))


"""thermodynamic temperature"""
T = 300
"""Boltzmann constant"""
kB = 0.0083144621
beta = 1.0 / (kB*T)
nMer = 3
nDOF = nMer * 3
system = Monomers(beta, nMer, 1.0, 10.0, np.zeros(3, dtype=float), 10.0, 10.0)
nBijectors = 5
bPermute = True
bijectors = []
permutations = []
"""
In order to fully permute all the dimension, I need to flatten out the
inner dimension so that tfb.Permute can work on it in MAF then I need
to reshape it for the log prob
"""
bijectors.append(tfb.Reshape(event_shape_in=[nMer, 3],
                             event_shape_out=[nDOF]))
for i in range(nBijectors):
    bijectors.append(tfb.Invert(tfb.MaskedAutoregressiveFlow(
        shift_and_log_scale_fn=tfb.masked_autoregressive_default_template(
            activation=tf.nn.sigmoid,
            hidden_layers=[nDOF]),
        unroll_loop=True)))
    if bPermute and i < nBijectors - 1:
        """Last bijector must not be permutation"""
        permutation = np.random.permutation(nDOF)
        bijectors.append(tfb.Permute(permutation=permutation, axis=-1))
        permutations.append(permutation)
bijectors.append(tfb.Reshape(event_shape_in=[nDOF], event_shape_out=[nMer, 3]))
"""The bijectors are applied from last to first -- need to reverse"""
flow = tfb.Chain(list(reversed(bijectors)))
baseDist = tfd.Independent(
    distribution=tfd.Uniform(low=tf.zeros([nMer, 3], dtype=tf.float32),
                             high=tf.ones([nMer, 3], dtype=tf.float32)),
    reinterpreted_batch_ndims=2)
dist = tfd.TransformedDistribution(
    distribution=baseDist,
    bijector=flow)
```
</p>
</details>

```python
lossReverse = tfp.vi.monte_carlo_csiszar_f_divergence(
    f=tfp.vi.kl_reverse,
    p_log_prob=system.log_prob,
    q=dist,
    num_draws=128)
optimizerReverse = tf.compat.v1.train.AdamOptimizer(1e-2)
trainReverse = optimizerReverse.minimize(lossReverse)

sess.run(tf.compat.v1.global_variables_initializer())

nStepTrainingReverse = int(2e4)
stepsReverse = []
iStep = 0
for i in range(nStepTrainingReverse):
    iLoss, _ = sess.run([lossReverse, trainReverse])
    stepsReverse.append(iStep)
    if i % 1000 == 0:
        print("Reverse: %10d%12.3e\n" % (i, iLoss))
    iStep += 1

```
I expect the eager execution is a bit slower but not 100x slower. Is there anything I am missing when migrating tensorflow probability code from 1.0 to 2.0?

**Comment by junpenglao** [comment]
Calling Eager function directly like this is extremely slow indeed, I think the very least is to wrap the optimized call in a function and decorated with tf.function. Try for example:
```python
def ELBOLoss():
  return tfp.vi.monte_carlo_variational_loss(
    system.log_prob,
    dist,
    discrepancy_fn=tfp.vi.kl_reverse,  # same as: Evidence Lower BOund
    sample_size=128)

optimizerReverse = tf.optimizers.Adam(1e-2)
@tf.function(autograph=False)
def train():
  optimizerReverse.minimize(ELBOLoss, var_list=dist.trainable_variables)
  return ELBOLoss()

nStepTrainingReverse = int(2e4)
stepsReverse = []
iStep = 0
for i in range(nStepTrainingReverse):
    stepsReverse.append(iStep)
    train()
    if i % 1000 == 0:
        print("Reverse: %10d\n" % (i))
    iStep += 1
```

**Comment by dynamicwebpaige** [comment]
🙂 Thank you for the answer, @junpenglao! @Char-Aznable, were you able to see similar performance to your TensorFlow 1.x implementation when using `@tf.function` in TF 2.0?

**Comment by junpenglao** [comment]
There might be other TF2.0 issues here - seems `dist.trainable_variables` is empty. @csuter do you know if `masked_autoregressive_default_template` works under TF2.0 yet?

**Comment by Char-Aznable** [comment]
@dynamicwebpaige I got the speed up with @junpenglao's suggestion (it's actually 2x faster than the TF v1 without eager execution) but with wrong answer. Indeed, `dist.trainable_variables` is empty and the model is not trained at all.

**Comment by junpenglao** [comment]
cc @brianwa84 @jvdillon for additional inputs.

**Comment by csuter** [comment]
MaskedAutoregressive bijector (and default template fn) have unit tests that run and pass in TF1 (graph) style, and TF2 (eager) style, so at a glance I don't know any reason why TF2 would be causing issues.

Is it any help to use the unit tests for this bijector as a guide to usage? https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/bijectors/masked_autoregressive_test.py

**Comment by junpenglao** [comment]
Hmm, seems the `trainable_variable` or initializing the bijector with Variable is never tested for MaskedAutoregressive bijector and default template fn? That could explain why it is passing but actually not gradient safe yet.

**Comment by csuter** [comment]
Ah ok, it looks like the tf.Module trainable_variable detection magic we rely on is not going to be able to detect the variables created inside the masked_autoregressive_default_template. This is a real bug, which we should fix. It also looks like the underlying `masked_dense` routine is using tf1.layers, which may not be related but is anachronistic and should also be updated.

**Comment by junpenglao** [comment]
Update: 
masked_autoregressive_default_template is now deprecated in https://github.com/tensorflow/probability/commit/b99d9647239b77159bb2d344570cc40c70e689ac 
You will need to use `tfb.AutoregressiveNetwork` instead

Below is a snippet:
```python
dims = 2
made = tfb.AutoregressiveNetwork(params=2, hidden_units=[10, 10])
bij = tfb.MaskedAutoregressiveFlow(
    shift_and_log_scale_fn=made)
maf = tfd.TransformedDistribution(
    distribution=tfd.Normal(loc=0., scale=1.),
    bijector=bij,
    event_shape=[dims])
maf.sample()
maf.trainable_variables
```

There are also some very informative doc string in regarding variable tracking: https://github.com/tensorflow/probability/blob/2065e03df67a35d62145b527ef0f41d4d38e61b0/tensorflow_probability/python/bijectors/masked_autoregressive.py#L197

I will close this for now but feel free to keep discussing especially if you are having difficult migrating to using `AutoregressiveNetwork`
